xbmc/windows/GUIMediaWindow.cpp

```diff
@@ -1488,19 +1488,170 @@ bool CGUIMediaWindow::OnPlayMedia(int iItem, const std::string &player)
  *
  * This function is called by OnClick()
  */
+class QueueMediaAsync : public CJob
+{
+public:
+  QueueMediaAsync(int iPlaylist, std::unique_ptr<CFileItemList> Localm_vecItems)
+  {
+    m_iPlaylist = iPlaylist;
+    m_Localm_vecItems.reset(new CFileItemList);
+    m_Localm_vecItems->Copy(*Localm_vecItems);
+  }
+
+  ~QueueMediaAsync(void) {}
+
+  bool DoWork(void)
+  {
+    bool ItemDetected = false;
+    int PlaylistSize = 0;
+    int Localm_vecItemsSize = m_Localm_vecItems->Size();
+
+    for ( int i = 0; i < Localm_vecItemsSize; i++ )
+    {
+      if (ShouldCancel(0, 0))
+        return false;
+
+      // Perform Playlist refresh every 50 items
+      if ( i % 50 == 0 )
+      {
+        CGUIMessage msg(GUI_MSG_PLAYLIST_CHANGED, 0, 0);
+        CServiceBroker::GetGUI()->GetWindowManager().SendMessage(msg);
+        Event.WaitMSec(100);
+      }
+
+      CFileItemPtr nItem = m_Localm_vecItems->Get(i);
+      PlaylistSize = CServiceBroker::GetPlaylistPlayer().GetPlaylist(m_iPlaylist).size();
```

Comment:
Move this  outside the loop. No need to obtain the same value on each iteration.

Reply:
The PlaylistSize could have changed in worker thread, cause the OnPlayAndQueueMedia routine could have added an item while worker is active.
e.g. whole "itemlist" has 1000 songs (in my video I use the genre selection as "itemlist")
Now I select one e.g. 10th song from Itemlist. This song is added immediately to playlist. Now worker triggers... 
first loop is "for ( int i = 0; i < Localm_vecItemsSize; i++ )"
i=0 so try adding first song but verify if not already added (10th song already added)
get 
PlaylistSize = CServiceBroker::GetPlaylistPlayer().GetPlaylist(m_iPlaylist).size();
second loop for verification:
for ( int j = i; j < PlaylistSize; j++ )
range 0 to 0 (one item (the 10th song) already in playlist so we check if item = playlistitem
and so on.

Now imagine following. we add 900th song. Worker starts and progress takes time.
On large playlists I can add another song  e.g. 950 via "OnPlayAndQueueMedia" (therefore is the signature verification).
now, e.g. worker is at position 500 but playlistsize is now 502. -> Playlistsize changed.
and so on...